### PR TITLE
app-project: Refactor NavLink to handle internal vs external route

### PR DESCRIPTION
**Pull request by goplayoutside3**
## Package
app-project

## Linked Issue and/or Talk Post
Related to https://github.com/zooniverse/front-end-monorepo/pull/6318
Links to external routes should not use `next/link`. PFE pages are considered external routes as well as fe-root. They're all completely separate web apps.

## Describe your changes
- Removed `disabled` state from NavLink. This prop is not used anywhere in combo with NavLink in app-project. There's no reason to keep it, and much more elegant solutions exist to display a disabled link or button as needed such as a [disabled PlainButton](https://zooniverse.github.io/front-end-monorepo/@zooniverse/react-components/index.html?path=/story/components-plainbutton--default&args=disabled:!true)).
- Added `externalLink` option to the `link` prop passed to shared component NavLInk. Rather than rip out NavLink from app-project components that shouldn't be using a `next/link` (there's a lot of them 😬 ), I refactored NavLink with logic to handle whether to use `next/link` depending on:
    ```
    link = {
      externalLink: true,
      href: 'etc',
      text: 'etc'
    }
    ```
    -  I also chose this solution because app-project has a shared component called ContentBox that can accept `linkProps` which get passed on to NavLink. I only found one case where this was actually happening (project stats), but in order to not disrupt the one-size-fits-all ContentBox behavior, I chose to adapt NavLink's code.

## How to Review

This is best tested by running locally in _production_ mode:
- Run [any project locally](https://docs.google.com/spreadsheets/d/1X9345vQ_xNX6BJf9_xwVCQCjxjbWWErGmV7Xe8WpHZo/edit?gid=0#gid=0) and check that the ProjectHeader links do client-side routing for project home, about, and classify. You'll see prefetch requests logged to the terminal when hovering over ProjectTitle (link to homepage), or About or Classify. Same on any project's About pages, prefetching happens on all of the links in the About Project sidebar.
- On any project's Team page, the username should still form the link-to-user correctly.
- Link to a project's stats page should also be formed correctly.

(You will still see one failed prefetch for `/talk` in the network tab because of the JoinInButton handled in https://github.com/zooniverse/front-end-monorepo/pull/6318).

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Bug Fix
- [ ] The PR creator has listed user actions to use when testing if bug is fixed
- [ ] The bug is fixed
- [ ] Unit tests are added or updated

### CODE REVIEW

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Package: `app-project`

This project changes how the links in the Nav and DropdownNav of the ProjectHeader

Screenshot: project header Nav (desktop/view view) and DropdownNav (mobile/narrow view)

image

image

- The links in the nav are now distinguished between "internal" and "external" links.
  - Internal links take users to FEM pages. External links take users to PFE pages.
  - <details><summary>Internal links use the Next.js Link component. External links don't.</summary> (IIUC, this is due to how Next.js's Links prefetch the linked pages, which only works if the linked pages are running on Next.js, i.e. it only makes sense for FEM pages.)</details>
- For app-project's Project Header, only **two links** are considered internal: the **About** link and the **Classify** link.

The update works, as far as I can tell.
- Testing with [a random test project on localhost in the production env,](https://local.zooniverse.org:3000/projects/darkeshard/test-project-2021?env=production) clicking on the links correctly take me to the expected pages. 👍 
- I see prefetch notices in the dev console, but am not familiar enough with Next.js to reliably replicate their success/fail scenarios.

❗ That said, I added some comments about how the current code implementation (i.e. always assuming _the first two_ items in the nav list are internal) can be brittle, and recommended an alternative (i.e. explicitly spell out which links are internal/external.)

### Status

I'm approving this PR since it works, but please consider my suggestion to make distinguishing the external/internal links a more robust process.

File: packages/app-project/src/components/ProjectHeader/components/DropdownNav/DropdownNav.js [hunk 1]
```diff
@@ -88,18 +88,21 @@ function DropdownNav({
...
     >
-      <Box as='ul' pad='0px'>
-        {navLinks.map(navLink => (
+      <Box as='ul' pad='0'>
+      {navLinks?.slice(0, 2).map(navLink => (
           <NavItem key={navLink.href} navLink={navLink} />
         ))}
+        {navLinks?.slice(2).map(navLink => (
+          <NavItem key={navLink.href} navLink={{ ...navLink, externalLink: true }} />
+        ))}
```
[the remark this answers is not in the document]
Comment (goplayoutside3) [reply packages/app-project/src/components/ProjectHeader/components/DropdownNav/DropdownNav.js:105]: Good point! I'll refactor before merging, thanks!

**Comment by goplayoutside3** [comment]
@shaunanoordin thanks for the review! I refactored the addition of `externalLink: true` to the useProjectNavigation hook.

To respond to a couple of your comments:
> Internal links take users to FEM pages. External links take users to PFE pages.

While this is true, we have three apps on the frontend. We often generalize app-project and app-root into just "FEM", but they are in fact separate apps. This is why app-root and app-project cannot use `next/link` in the ZooHeader component despite the homepage, About, and Get Involved all existing in "FEM". ZooHeader doesn't know whether its parent is app-root or app-project, so I'm just writing this out as an example of "Internal links take users to a page in the _same application_. External links take users to a _separate application_".

> I see prefetch notices in the dev console, but am not familiar enough with Next.js to reliably replicate their success/fail scenarios. 

You can also see prefetch notices in the network tab. When prefetch attempts a non-internal page, you'll see a 404 for that request in both the network tab and browser console (when run in production mode!). This branch is updated with https://github.com/zooniverse/front-end-monorepo/pull/6318, so now there are no 404's at all 👍

**Comment by eatyourgreens** [comment]
I’m not sure about the app router, but `next/link` in the pages router renders a regular `<a>` tag if the target path doesn’t exist in the Next.js app. 

We had some problems ages ago because the project app had stub paths for `/talk`. Removing the stubs fixed all the broken `Link` components. 

Of course, Vercel may have broken the `Link` component again, since then.

**Pull request merged**